link/uprobe.go

```diff
@@ -67,42 +67,66 @@ func OpenExecutable(path string) (*Executable, error) {
 		return nil, fmt.Errorf("parse ELF file: %w", err)
 	}
 
-	var ex = Executable{
+	ex := Executable{
 		path:    path,
-		symbols: make(map[string]elf.Symbol),
+		offsets: make(map[string]uint64),
 	}
-	if err := ex.addSymbols(se.Symbols); err != nil {
+
+	if err := ex.load(se, se.Symbols); err != nil {
 		return nil, err
 	}
 
-	if err := ex.addSymbols(se.DynamicSymbols); err != nil {
+	if err := ex.load(se, se.DynamicSymbols); err != nil {
 		return nil, err
 	}
 
 	return &ex, nil
 }
 
-func (ex *Executable) addSymbols(f func() ([]elf.Symbol, error)) error {
-	// elf.Symbols and elf.DynamicSymbols return ErrNoSymbols if the section is not found.
-	syms, err := f()
+func (ex *Executable) load(f *internal.SafeELFFile, fun func() ([]elf.Symbol, error)) error {
+	// Returns ErrNoSymbols if the section is not found.
+	syms, err := fun()
 	if err != nil && !errors.Is(err, elf.ErrNoSymbols) {
 		return err
 	}
+
 	for _, s := range syms {
 		if elf.ST_TYPE(s.Info) != elf.STT_FUNC {
 			// Symbol not associated with a function or other executable code.
 			continue
 		}
-		ex.symbols[s.Name] = s
+
+		off := s.Value
+		if f.Type == elf.ET_EXEC || f.Type == elf.ET_DYN {
+			for _, prog := range f.Progs {
+				if prog.Type != elf.PT_LOAD || (prog.ProgHeader.Flags&elf.PF_X) == 0 {
+					continue
+				}
+
+				if prog.Vaddr <= s.Value && s.Value < (prog.Vaddr+prog.Memsz) {
+					off = s.Value - prog.Vaddr + prog.Off
+				}
+			}
+		}
+
+		ex.offsets[s.Name] = off
 	}
+
 	return nil
 }
 
-func (ex *Executable) symbol(symbol string) (*elf.Symbol, error) {
-	if s, ok := ex.symbols[symbol]; ok {
-		return &s, nil
+func (ex *Executable) offset(symbol string) (uint64, error) {
+	if off, ok := ex.offsets[symbol]; ok {
+		// Symbols with location 0 from section undef are shared library calls and
```

Comment:
Can you break this change into a separate commit and explain why moving it makes sense?

Reply:
My reasoning is: since we have to do operations on the offsets, we either store a fake `elf.Symbol` with the modified offset value or store the offset values directly. I decided to go for the latter and move this check where we get the value.
Since I don't have an `elf.Symbol` to check anymore, I am now assuming that if the offset value is 0, we have an external symbol. (I am not 100% sure about this assumption so the error text may be misleading here)

Let me know if you have a better idea